### PR TITLE
Final version: Ed25519 with EdDSA compatibility

### PR DESCRIPTION
**Pull request by nikkolasg**
Finally after some talks, this is the right PR where we do the compatibility with EdDSA. I will delete #68 #69 #71 .

### CODE REVIEW

File: nist/int.go [hunk 1]
```diff
@@ -37,8 +39,9 @@ var two = big.NewInt(2)
 // whose target is assumed never to change.
 //
 type Int struct {
-	V big.Int  // Integer value from 0 through M-1
-	M *big.Int // Modulus for finite field arithmetic
+	V          big.Int          // Integer value from 0 through M-1
+	M          *big.Int         // Modulus for finite field arithmetic
+	endianness binary.ByteOrder // Endianness considered for this int
```
Comment (bford) [inline nist/int.go]: Maybe rename the element 'bo' or 'byteOrder' since that seems more self-explanatory (and more like a word) than "endianness"?

And perhaps just make it public (e.g., 'BO'), since the other elements are already public and I don't see any obvious harm/risk if this one is public too?

Finally, I'm not totally sure 'binary.ByteOrder' is the appropriate type to use here, since that's a whole interface-pointer that could in principle point to any object implementation that interface, but here the semantics are that we *need* it to be 'BigEndian' or 'LittleEndian' and nothing else.  i.e., if the caller creates their own 'myLittleEndian' object satisfying the ByteOrder interface and behaving exactly like LittleEndian, but a different pointer from LittleEndian, things will break, which violates the way interfaces are "supposed" to work.  So here I think the cleaner approach would actually be to declare a local ByteOrder type, e.g.:

type ByteOrder bool
const (
   LittleEndian ByteOrder = false
   BigEndian ByteOrder = true
)

...and just use that.

Comment (nikkolasg) [inline nist/int.go]: All right for the public. I also hesitated creating our own type but then I chose the *standard* way. But since as you noted it might break the thing, I now think that's also best to declare our own...

File: nist/int.go [hunk 2]
```diff
@@ -314,21 +313,17 @@ func (i *Int) UnmarshalBinary(buf []byte) error {
 	if len(buf) != i.MarshalSize() {
 		return errors.New("Int.Decode: wrong size buffer")
 	}
-	i.V.SetBytes(buf)
+	if i.endianness == binary.LittleEndian {
+		i.SetLittleEndian(buf)
+	} else {
+		i.V.SetBytes(buf)
```
Comment (bford) [inline nist/int.go]: SetBytes should be sensitive to the current byte-order, so this if-else shouldn't be needed here; it should happen in SetBytes instead.

File: nist/int.go [hunk 3]
```diff
@@ -346,6 +341,21 @@ func (i *Int) BigEndian(min, max int) []byte {
...
 
+// Set value to a number represented by a byte string.
+// Endianness depends on the endianess set in i.
+func (i *Int) SetBytes(a []byte) *Int {
+	i.V.SetBytes(a).Mod(&i.V, i.M)
+	return i
+}
+
+// Set value to a number represented in a little-endian byte string.
+func (i *Int) SetLittleEndian(a []byte) *Int {
```
Comment (bford) [inline nist/int.go]: This shouldn't be needed anymore, or at least doesn't need to be public; it should get merged into SetBytes right?

File: nist/int.go [hunk 4]
```diff
@@ -416,3 +434,14 @@ func (i *Int) HideDecode(buf []byte) {
 	i.V.SetBytes(buf)
 	i.V.Mod(&i.V, i.M)
 }
+
+// SetEndianess will set the endianness of the value to end.
+func (i *Int) SetEndianness(end binary.ByteOrder) {
```
Comment (bford) [inline nist/int.go]: Rename to SetByteOrder?

File: nist/int.go (hunk 3 above)
Comment (nikkolasg) [inline nist/int.go]: I also think that this should not be needed anymore but when I asked you you said "let's keep it" .. :) I'm all for removing that !

File: nist/int.go (hunk 4 above)
Comment (nikkolasg) [inline nist/int.go]: Since I'm going for a Public field as suggested above, we can even remove that method

File: nist/int.go (hunk 2 above)
Comment (nikkolasg) [inline nist/int.go]: Well, if `MarshalBinary` returns a slice in LittleEndian format, then `UnmarshalBinary` on the same `Int` should expect a slice in LittleEndian format ... I thought it was the whole point of the thing,  to be able to use  `Secret` which `Marshal/Unmarshal` in LittleEndian format... ?

File: random/rand.go [hunk 5]
```diff
@@ -86,6 +86,24 @@ func Bytes(n int, rand cipher.Stream) []byte {
 	return b
 }
 
+// NonZeroBytes calls Bytes as long as it gets a slice full of '0's.
+// This is needed when using suite.Cipher(abstract.NoKey)
+// because the first 6 iterations returns 0000...000 as
+// bytes for edwards & ed25519 cipher.
+// XXX Issue reported in https://github.com/dedis/crypto/issues/70
+func NonZeroBytes(n int, rand cipher.Stream) []byte {
```
Comment (nikkolasg) [inline random/rand.go]: I've put it here but I'm still hesitating whether this should go into `util` or stg like that . At least two packages will need this function.

File: nist/int.go (hunk 3 above)
Comment (bford) [inline nist/int.go]: I probably meant that I want to keep the *ability* to get little-endian behavior out of nist.Int - but that doesn't mean we need to keep the SetLittleEndian *method*. :)

File: nist/int.go (hunk 4 above)
Comment (bford) [inline nist/int.go]: Fine with me.  No real need for getters/setters if there's no important internal abstraction to be protected.

File: random/rand.go (hunk 5 above)
Comment (ineiti) [inline random/rand.go]: So there is no way of having the ciphers behaving in a non-zero way?

File: abstract/group.go [hunk 6]
```diff
@@ -51,6 +51,9 @@ type Secret interface {
 
 	// Set to a fresh random or pseudo-random secret
 	Pick(rand cipher.Stream) Secret
+
+	// SetBytes will take bytes and create a secret out of it
+	SetBytes([]byte) Secret
```
Comment (ineiti) [inline abstract/group.go]: Will we also have a `Bytes()`-method?

File: ed25519/suite.go [hunk 7]
```diff
@@ -37,8 +42,32 @@ func (s *suiteEd25519) New(t reflect.Type) interface{} {
...
+
 // Ciphersuite based on AES-128, SHA-256, and the Ed25519 curve.
 func NewAES128SHA256Ed25519(fullGroup bool) abstract.Suite {
 	suite := new(suiteEd25519)
 	return suite
 }
+
+// NewEd25519Scalar returns the scalar derived from the stream and the right
+// half of the hash during the derivation of the scalar.
+func NewEd25519Scalar(stream cipher.Stream) ([]byte, []byte) {
```
Comment (ineiti) [inline ed25519/suite.go]: Will this also be used elsewhere? If no, I'd include it in `NewKey`.

File: abstract/group.go (hunk 6 above)
Comment (nikkolasg) [inline abstract/group.go]: 1) There's no immediate reason to put it in this PR since it won't be used (But still can!)
2) Do you have the exact functionality of this function in mind ? 
 * `Marshal/Unmarshal` will give/read the representation of a secret correctly formatted and everything
 * `SetBytes` is supposed to accept any bytes input and squeeze that into a scalar. So I guess `Bytes()` is supposed to give back the original slice (1) ? Or does it returns the same thing as `MarshalBinary` (2) ? 
IF (1), then we need to keep track of an original slice alongside with the scalar (nist.Int).
IF (2), then there's absolutely no difference between `MarshalBinary` and `Bytes` 

I'm not so sure about both options, so what do you have in mind for the `Bytes()` function...

File: random/rand.go (hunk 5 above)
Comment (nikkolasg) [inline random/rand.go]: Yes there's surely a way since that is supposed to be their default behavior, return always random looking string, but for the moment there's a bug ( #70  ) so ...

File: eddsa/eddsa.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	Secret abstract.Secret
+	// Public is the corresponding public key
+	Public abstract.Point
+}
+
+// NewEdDSAKey will return a freshly generated key pair to use for generating
+// EdDSA signatures.
+// If stream == nil, it will take the random.Stream.
+func NewEdDSA(stream cipher.Stream) *EdDSA {
+	if stream == nil {
```
Comment (ineiti) [inline eddsa/eddsa.go]: Correcting myself: here you could use the `NewEd25519Scalar`

File: eddsa/eddsa.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	secret := suite.Secret().SetBytes(scalar[:32])
+	public := suite.Point().Mul(nil, secret)
+
+	return &EdDSA{buffer, scalar[32:], secret, public}
+}
+
+// Prefix returns the Prefix as being the right part of
+// the hashed seed
+func (e *EdDSA) Prefix() []byte {
+	c := make([]byte, len(e.seed))
```
Comment (ineiti) [inline eddsa/eddsa.go]: That would be `e.prefix`, no?

File: eddsa/eddsa.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+	buffer := random.NonZeroBytes(32, stream)
+	scalar := sha512.Sum512(buffer)
+	scalar[0] &= 0xf8
+	scalar[31] &= 0x3f
+	scalar[31] |= 0x40
+
+	secret := suite.Secret().SetBytes(scalar[:32])
+	public := suite.Point().Mul(nil, secret)
+
+	return &EdDSA{buffer, scalar[32:], secret, public}
```
Comment (ineiti) [inline eddsa/eddsa.go]: as the variable-names are different from the field-names of `EdDSA`, I'd do the `Seed: scalar[32:]`-type initialisation of the structure.

File: eddsa/eddsa.go (hunk 8 above)
Comment (ineiti) [inline eddsa/eddsa.go]: Oups - just saw that `NewEd25519Scalar` doesn't return the seed.

File: eddsa/eddsa.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+	copy(eddsa, e.seed)
+	copy(eddsa[32:], pBuff)
+	return eddsa, nil
+}
+
+// EdDSASign will return a EdDSA signature using Ed25519. The secret must be
+// Picked() so it can use the prefix, the right part of the hashing done in
+// Pick.
+// Takes
+// - suite is the suite to use
```
Comment (ineiti) [inline eddsa/eddsa.go]: Update comment.

File: eddsa/eddsa.go [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+	hash.Write(msg)
+
+	// deterministic random secret and its commit
+	r := suite.Secret().SetBytes(hash.Sum(nil))
+	R := suite.Point().Mul(nil, r)
+
+	// challenge
+	// H( R || Public || Msg)
+	hash.Reset()
+	Rbuff, err := R.MarshalBinary()
```
Comment (ineiti) [inline eddsa/eddsa.go]: Shouldn't we replace this with `R.Bytes()`? I thought `MarshalBinary` might compress the integer if the size of it is lower, no?

File: eddsa/eddsa.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+	if err != nil {
+		return nil, err
+	}
+	Abuff, err := e.Public.MarshalBinary()
```
Comment (ineiti) [inline eddsa/eddsa.go]: Question: does `MarshalBinary` follow the draft?

File: eddsa/eddsa.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+	// return R || s
+	var sig [64]byte
+	copy(sig[:], Rbuff)
+	copy(sig[32:], sBuff)
+
+	return sig[:], nil
+}
+
+// Verify verify a signature issued by EdDSA.Sign
```
Comment (ineiti) [inline eddsa/eddsa.go]: Typo

File: eddsa/eddsa.go (hunk 12 above)
Comment (nikkolasg) [inline eddsa/eddsa.go]: R is a point so `MarshalBinary` is the correct representation of a point.

File: eddsa/eddsa.go (hunk 13 above)
Comment (nikkolasg) [inline eddsa/eddsa.go]: Yes

File: eddsa/eddsa_test.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+// Comparing our implementation with the test vectors of the RFC
+func TestEdDSASigning(t *testing.T) {
+	for _, vec := range EdDSATestVectors {
+		seed, err := hex.DecodeString(vec.private)
+		assert.Nil(t, err)
+
+		stream := util.NewConstantStream(seed)
+
+		ed := NewEdDSA(stream)
```
Comment (ineiti) [inline eddsa/eddsa_test.go]: I think about what we discussed with Bryan and whether it's a good idea to use a stream for initialization when in reality we want it to be set to a given byte-slice. In crypto, `stream` is mostly used for random-streams.

File: eddsa/eddsa.go (hunk 8 above)
Comment (nikkolasg) [inline eddsa/eddsa.go]: You just write comment faster than I can code, I was removing it when you commented out

File: abstract/group.go (hunk 6 above)
Comment (ineiti) [inline abstract/group.go]: Isn't `MarshalBinary` compressing the integer if it's value allows it? I thought that:
- (Un)MarshalBinary are used to store things in a file - variable length, might have indexes and other stuff
- (Set)Bytes work directly on the value of the secret and always take/return the same length

File: eddsa/eddsa.go (hunk 8 above)
Comment (ineiti) [inline eddsa/eddsa.go]: Sorry...

File: eddsa/eddsa_test.go (hunk 15 above)
Comment (nikkolasg) [inline eddsa/eddsa_test.go]: > a stream for initialization when in reality we want it to be set to a given byte-slice

In **reality** (*not* in testing), we want that to be a *stream* so we get different bytes each time we call it. In **testing**, we want that to be a fixed slice of *bytes* so we know the expected result. For EdDSA, in **reality** we don't want to rely on a slice of bytes !

File: eddsa/eddsa.go [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+	}
+
+	hash.Write(Rbuff)
+	hash.Write(Abuff)
+	hash.Write(msg)
+
+	h := suite.Secret().SetBytes(hash.Sum(nil))
+
+	// response
+	// s = r  h * secret
```
Comment (ineiti) [inline eddsa/eddsa.go]: Typo: `s = r + h * s`

File: eddsa/eddsa.go [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+	s := suite.Secret().Mul(e.Secret, h)
+	s.Add(r, s)
+
+	sBuff, err := s.MarshalBinary()
```
Comment (ineiti) [inline eddsa/eddsa.go]: *The return of MarshalBinary/Bytes* - if it turns out to be correct that `MarshalBinary` on a secret might return fewer bytes, we need to use `Bytes()` here.

File: abstract/group.go (hunk 6 above)
Comment (ineiti) [inline abstract/group.go]: Just if somebody else thinks like me: you're wrong ;) `(Un)MarshalBinary` is fixed-size for `secret`s and `point`s.

Comment (bford) [inline abstract/group.go]: Wait, what?  From what I'm reading, it sounds like you're saying the exact opposite of what I thought the semantics should be. :)

- MarshalBinary() should ALWAYS produce exactly MarshalSize() bytes, independent of the value of the integer being marshaled.  That's why it has the code above to pad the return from i.V.Bytes() if it is too short.

- Bytes() should probably behave the same way as big.Int.Bytes(), namely returning a byte-slice just large enough to hold the actual integer being encoded.  

In other words, SetBytes() takes a variable-length integer of any size as input, and Bytes() produces a variable-length return.  UnmarshalBinary() always consumes *exactly* MarshalSize() bytes, and MarshalBinary() always produces *exactly* MarshalSize() bytes.

So the implementation of Bytes() in this module should just call i.v.Bytes(), then reverse the slice if the caller wants little-endian.

Make sense?

File: ed25519/suite.go (hunk 7 above)
Comment (bford) [inline ed25519/suite.go]: Agreed (or just make it private to the module).

Also, no need for function names in this module to contain "Ed25519" when they're in a module called "ed25519"!

Comment (nikkolasg) [inline ed25519/suite.go]: It's already removed since > 10mn now , reload your page! :D

File: random/rand.go (hunk 5 above)
Comment (ineiti) [inline random/rand.go]: I'd leave it here, as that method is only used till the bug will (eventually) be fixed.

File: eddsa/eddsa.go [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
```
Comment (bford) [inline eddsa/eddsa.go]: No, exactly the opposite, as discussed above.  MarshalBinary() always returns a fixed-size slice, and I'm pretty sure is what you want here.

File: abstract/group.go (hunk 6 above)
Comment (nikkolasg) [inline abstract/group.go]: > In other words, SetBytes() takes a variable-length integer of any size as input, and Bytes() produces a variable-length return

So you want to keep a reference to the original slice then in nist.Int ... ? For me it does not make sense to keep twice the information where it's not needed. If a user called `SetBytes` it can still keep the original reference to the slice, we should not need to duplicate the information (the original slice and the *squeezed* version)

File: nist/int.go [hunk 19]
```diff
@@ -37,8 +45,9 @@ var two = big.NewInt(2)
 // whose target is assumed never to change.
 //
 type Int struct {
-	V big.Int  // Integer value from 0 through M-1
-	M *big.Int // Modulus for finite field arithmetic
+	V  big.Int   // Integer value from 0 through M-1
+	M  *big.Int  // Modulus for finite field arithmetic
+	Bo ByteOrder // Endianness considered for this int
```
Comment (bford) [inline nist/int.go]: Shouldn't it be BO rather than Bo, since it's an abbreviation and not a single word "bo"?

File: abstract/group.go (hunk 6 above)
Comment (ineiti) [inline abstract/group.go]: No need to keep the original around. You can just squeeze whatever int you have to remove leading zeroes.

File: nist/int.go [hunk 20]
```diff
@@ -299,10 +298,16 @@ func (i *Int) MarshalSize() int {
 func (i *Int) MarshalBinary() ([]byte, error) {
 	l := i.MarshalSize()
 	b := i.V.Bytes() // may be shorter than l
-	if ofs := l - len(b); ofs != 0 {
+	offset := l - len(b)
+
+	if i.Bo == LittleEndian {
+		return i.LittleEndian(l, l), nil
```
Comment (bford) [inline nist/int.go]: Ooh, what is the reason for the LittleEndian() method to exist on this object.  Looks unnecessary and kinda hairy and gross to me.  Why not just delete it and instead:

1. Encode into a big-endian byte-slice, handling the offset correctly for big-endian as the code below does.
2. *after* that (before line 307/312 below), test for LittleEndian and conditionally to a util.Reverse on the encoded slice.

Wouldn't that be way simpler?  Am I missing anything?

Comment (nikkolasg) [inline nist/int.go]: I also think it's not necessary now, but I understood you wanted to keep them, I again may have misunderstood. For me `LittleEndian` &  `BigEndian`  should be removed, does that seem OK ?

File: nist/int.go [hunk 21]
```diff
@@ -346,6 +343,18 @@ func (i *Int) BigEndian(min, max int) []byte {
 	return buf
 }
 
+// Set value to a number represented by a byte string.
+// Endianness depends on the endianess set in i.
+func (i *Int) SetBytes(a []byte) abstract.Secret {
+	var buff = a
+	if i.Bo == LittleEndian {
+		buff = make([]byte, len(a))
+		util.Reverse(buff, a)
```
Comment (bford) [inline nist/int.go]: I just realized that util.Reverse() should really have append()-like or Hash()-like semantics, allocating or extending the destination slice automatically if it is too short.  That simplify this and probably other calling code - e.g., here, this would become simply 'a = util.Reverse(nil, a)', and you could delete the 'buff' variable entirely.

Comment (bford) [inline nist/int.go]: Note that the trivial way to fix util.Reverse() to have this behavior is for it simply to call util.Grow() on its first argument.

Comment (nikkolasg) [inline nist/int.go]: Yep, could simplify things. Doing it here..

Comment (nikkolasg) [inline nist/int.go]: > Note that the trivial way to fix util.Reverse() to have this behavior is for it simply to call util.Grow() on its first argument.

Well not quite because `util.Grow` still expect a non-nil slice ...

File: random/rand.go (hunk 5 above)
Comment (bford) [inline random/rand.go]: Sounds good.  In the meantime you could at least make it non-public (assuming it's only used internally within this module).

File: util/stream.go [hunk 22]
```diff
@@ -0,0 +1,19 @@
+package util
+
+import "crypto/cipher"
+
+// ConstantStream is a cipher.Stream which always returns
+// the same value.
+type ConstantStream struct {
```
Comment (bford) [inline util/stream.go]: Why should this type be public?

File: random/rand.go (hunk 5 above)
Comment (nikkolasg) [inline random/rand.go]: No it's not, it's used by `ed25519` and `edwards`

File: util/stream.go [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+	seed []byte
+}
+
+// NewConstantStream returns a ConstantStream seeded with buff
+func NewConstantStream(buff []byte) cipher.Stream {
```
Comment (bford) [inline util/stream.go]: Just rename this function to ConstantStream() and make it return a value of a hidden type, since there's no reason for the caller to be aware of anything but the cipher.Stream interface to it.

File: util/stream.go (hunk 22 above)
Comment (nikkolasg) [inline util/stream.go]: Because since users are going to have a reference on it, they may have to understand what's it about and for the doc to show the comments it must be public ... ?

File: abstract/group.go (hunk 6 above)
Comment (nikkolasg) [inline abstract/group.go]: Well I don't get it then how you can return the original variable length slice byte without storing a reference to it ...

File: util/stream.go (hunk 22 above)
Comment (bford) [inline util/stream.go]: Just put the relevant comments on the function that produces the object of the hidden type!  No need to have an unnecessary public type just in order to hang comments off of!

Comment (nikkolasg) [inline util/stream.go]: It's already done, but github still shows it .. .><

File: abstract/group.go (hunk 6 above)
Comment (bford) [inline abstract/group.go]: No, I'm pretty sure big.Int.Bytes() just returns a *new* slice into which it copies a representation of the integer - and that's what nist.Int.Bytes() should do as well.  Note that big.Int has a Words() function that actually returns the internal-representation slice, but that's a slice of (32-bit?) words instead of bytes for efficiency, and it's only intended for use in performance-critical extensions to big.Int that really need to manipulate the internal state efficiently.  That's definitely not what Bytes() is for.  Just create a new slice, encode into it, and return it.

Comment (nikkolasg) [inline abstract/group.go]: Yes ok,  I get it this way. For me it still does not make much sense to for `SetBytes` and `Bytes` to not use the same slice. Example with `Un/Marshal` we can do :
```
a := sec.MarshalBinary()
sec.UnmarshalBinary(a)
a2 := sec.MarshalBinary()
a == a2
```
with `Bytes`:
```
secret.SetBytes(a)
a2 := secret.Bytes()
a != a2
```
But ok, if that makes sense for most of you then I abide :)

Comment (bford) [inline abstract/group.go]: Oooh - I'm pretty sure you really don't want to depend on being able to do a == a2 even in the MarshalBinary case!  Is there code that is doing that?  Sounds like a really bad idea to me - whether it works would definitely depend on which particular implementation of Secret or Point we're looking at.  For Point.MarshalBinary(), for example, I think special encoding is always going to be required, which will always allocate a new byte-slice.  If that's not happening for ed25519.Secret, then that might be a bug that needs to be fixed.

Comment (nikkolasg) [inline abstract/group.go]: Well, for the `(Un)Marshal` case, the snippet I posted above seems very normal and required to me, otherwise we get different representations of a secret ? 
And I've made the changes for the `Bytes()` now.

Comment (bford) [inline abstract/group.go]: Oh wait - by "a == a2" you meant a comparison of the *contents* of the slices, not a comparison of the slice-pointers?  (Because syntactically, "a == a2" in Go does *not* do a slice contents-comparison, if that's what you meant.)

If you did mean slice content-comparison, then yes, Marshal/Unmarshal can and should be expected to differ from SetBytes/Bytes in certain semantic subtleties.  

Marshal then Unmarshal then Marshal again should always produce the same thing the second time, since its purpose is to encode the internal state into its appropriate fixed-length "native" representation.  Similarly, if you do:

a := sec.Bytes()
sec.SetBytes(a)
a2 := sec.Bytes()

...then a byte wise comparison of a and a2 should always return true, even if the length of a (and a2) may vary depending on the specific integer contained in 'sec'.

However, UnmarshalBinary definitely should differ from SetBytes in that there are some inputs to SetBytes (e.g., 512-bit integers) for which a subsequent Bytes() won't yield the same thing.  Whereas if you do sec.UnmarshalBinary(a), then sec.MarshalBinary(), you should always get back a *if* a was actually a valid encoding of something.  (Note that sec.UnmarshalBinary() might return an error if not.)

Comment (nikkolasg) [inline abstract/group.go]: Yes sorry I meant in my *pseudo code* a byte content checking. And what you wrote is clear in my head but I'm just not so a fan of having this behavior of  `(Set)Bytes()`. Using `big.Int` it stays consistent: https://play.golang.org/p/pm3eAfoKYg
But anyway, as i'm clearly alone in my opinion, I already did the changes and that's fine, I'll still be able to sleep at night ;)

Comment (bford) [inline abstract/group.go]: But big.Int has quite different semantics from nist.Int.  big.Int is supposed to store an *arbitrary* big-integer.  nist.Int is supposed to store a *modular* integer (i.e., an element of a modular integer group or a scalar secret/exponent).  Because of this difference, big.Int by definition never needs to "decide" what to do if an integer supplied to SetBytes is "too big" because no integer is "too big".  nist.Int by contrast does need to decide what to do with a too-big integer, and the most convenient/useful semantics seems to apply the appropriate modulus.

Actually, perhaps some of your discomfort comes from the fact that "nist.Int" is probably misnamed.  For one thing, it should probably not be part of the 'nist' package at all (especially since that package probably should not even be part of the v1.0-stable version of crypto we're preparing), but probably elsewhere - I just haven't quite figured out where yet.  For another thing, perhaps it should not just be named 'Int' but instead perhaps 'ModInt' or something like that, to make it clear that it's not just a "different version of" big.Int but in fact does something else (which is true).

Comment (bford) [inline abstract/group.go]: Or when we rename abstract.Secret to abstract.Scalar, maybe nist.Int should become abstract.IntScalar, representing a "standard/default implementation" of the abstract.Scalar interface?

Comment (nikkolasg) [inline abstract/group.go]: I totally understand all the **semantics** and **practical** differences between nist.Int and big.Int but even though, from an **API perspective**, I should have said, it still *feels wrong*.  `(Set)Bytes` feels like it's a getter/setter interface while it's not because it introduces change (modulo) in the middle. Moreover, the `Bytes()` functionality can be questioned regarding its utility because it can be emulated using `MarshalBinary` and then cut off the padding.

Well, anyway, it's an endless loop and I think we should move on, so I accept this solution without any regret ;)

Regarding the naming, yes it would be a good idea to rename it `ModInt` or `IntMod` stg along those lines...

File: random/rand.go (hunk 5 above)
Comment (Liamsi) [inline random/rand.go]: If you still wish to hide it from outside, you could create a `internal` package on the top-level (which can safely be deleted later)

File: abstract/suite.go [hunk 24]
```diff
@@ -39,6 +40,10 @@ type Suite interface {
 	// Generic constructor to instantiate any abstract interface type
 	// supported by this suite: at least Cipher, Hash, Point, Secret.
 	Constructor
+
+	// NewKey returns a freshly generated private key from the cipher stream.
+	// If cipher == nil, it uses random.Stream.
+	NewKey(cipher.Stream) Secret
 }
```
Comment (Liamsi) [inline abstract/suite.go]: Question: The idea to have a `GenKeyPair` method in each suite is obsolete?

File: eddsa/eddsa.go [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+	}
+
+	hash.Write(Rbuff)
+	hash.Write(Abuff)
+	hash.Write(msg)
+
+	h := suite.Secret().SetBytes(hash.Sum(nil))
+
+	// response
+	// s = r  h * s
```
Comment (Liamsi) [inline eddsa/eddsa.go]: Isn't sth. missing in the comment? (A `+`)?

File: eddsa/eddsa.go [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+	if err != nil {
+		return err
+	}
+	hash := sha512.New()
+	hash.Write(sig[:32])
+	hash.Write(Pbuff)
+	hash.Write(msg)
+
+	h := suite.Secret().SetBytes(hash.Sum(nil))
+	// reconstruct S == k*A  R
```
Comment (Liamsi) [inline eddsa/eddsa.go]: Do you skip `+` on purpose? It's only an internal comment.

File: abstract/group.go (hunk 6 above)
Comment (Liamsi) [inline abstract/group.go]: The `Bytes()` function is implemented but not in the interface... (like `SetBytes(...)` which is)

**Comment by Liamsi** [comment]
Besides some mini-glitches: looks OK to me.

File: abstract/suite.go (hunk 24 above)
Comment (nikkolasg) [inline abstract/suite.go]: I went for the simplest. We could also have done `NewKeyPair(cipher.Stream) config.KeyPair` but config package already uses abstract, so that would mean to move the `config.KeyPair` into abstract which do not seem something that has to be done in this PR... :)
